sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/impl/parser/TestTableProvider.java

```diff
@@ -17,38 +17,165 @@
  */
 package org.apache.beam.sdk.extensions.sql.impl.parser;
 
-import java.util.HashMap;
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.List;
 import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.coders.Coder;
 import org.apache.beam.sdk.extensions.sql.BeamSqlTable;
+import org.apache.beam.sdk.extensions.sql.impl.schema.BeamIOType;
 import org.apache.beam.sdk.extensions.sql.meta.Table;
 import org.apache.beam.sdk.extensions.sql.meta.provider.TableProvider;
+import org.apache.beam.sdk.extensions.sql.mock.MockedBoundedTable;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+import org.apache.beam.sdk.values.POutput;
+import org.apache.beam.sdk.values.Row;
 
-/** Test in-memory table provider for use in tests. */
+/**
+ * Test in-memory table provider for use in tests.
+ *
+ * <p>Keeps global state and tracks class instances. Works only in DirectRunner.
+ *
+ * <p>Similar to {@link MockedBoundedTable} with more concurrency support.
+ */
 public class TestTableProvider implements TableProvider {
-  private Map<String, Table> tables = new HashMap<>();
+  static final Map<Long, Map<String, TableWithRows>> GLOBAL_TABLES = new ConcurrentHashMap<>();
```

Comment:
Could you elaborate why convert tables to static and change `BeamSqlTable` to `synchronized`?

It seems like it is changed for multi-threading?

Reply:
This is the same idea as in EnumerableConverter. We keep accumulating data in a global state. The state is keyed by the instance ID of the table provider, which creates tables. Few queries can run in parallel against different tables, so we need something like this